code_sync/code_sync.py

```diff
@@ -2,39 +2,186 @@
           '{target}:{remote_dir}" {local_dir}'
 
 epilog_str = '''
-Example for connecting to LeoMed:
-    code_sync --local_dir mylocaldir/ --remote_dir myremotedir/ --target medinfmk --port 2222\n
+EXAMPLE USAGE
+Register a project:
+    code_sync --register <project>
+
+code_sync a registered project:
+    code_sync <project>
+
+List all projects registered to code_sync:
+    code_sync --list
+
+Run code_sync with specific parameters:
+    code_sync --local_dir <mylocaldir/> --remote_dir <myremotedir/> --target <ssh_remote> --port 2222\n
 
 '''
 
+CONFIG_FILE_NAME = '.code_sync'
+
 
 def code_sync(local_dir, remote_dir, target, port=22):
     # clean up slashes
     local_dir = os.path.join(local_dir, '')
     remote_dir = os.path.join(remote_dir, '')
 
-    # subprocess.call()
+    print("Starting code_sync between {} and {}:{} ...".format(local_dir, target, remote_dir))
+    print('(^C to quit)')
     cmd = cmd_str.format(local_dir=local_dir, remote_dir=remote_dir, target=target, port=port)
     subprocess.call(cmd, shell=True)
 
 
+def load_config() -> Dict:
+    """
+    Load the code_sync config file. Create a blank one if no file exists.
+
+    Returns:
+        The config loaded from the file.
+    """
+
+    create_config_if_not_exists()
+
+    config_file_path = Path(Path.home(), CONFIG_FILE_NAME)
+    with open(config_file_path, 'r') as f:
+        config = yaml.safe_load(f)
+    # if config is empty, return an empty dictionary (not None)
+    if config is None:
+        config = {}
+    return config
+
+
+def init_config() -> None:
+    """Create an empty config file."""
+    config_path = Path(Path.home(), CONFIG_FILE_NAME)
+    open(config_path.__str__(), 'x').close()
+
+
+def create_config_if_not_exists() -> None:
+    """Create the code_sync config if it does not already exist."""
+    config_file_path = Path(Path.home(), CONFIG_FILE_NAME)
+    if not config_file_path.exists():
+        init_config()
+
+
+def register_project(project: Union[str, None]) -> None:
+    """
+    Register a project to the code_sync config.
+
+    Args:
+        project: The name of the project to register. If None is provided, the user will be asked for a project name.
+
+    Returns:
+        None. The result is saved to the code_sync config.
+
+    Raises:
+        ValueError if there is already a registered project with the given name.
+
+    """
+    if project is None:
+        project = input('Project name: ')
+    else:
+        print("Registering new project '{}'".format(project))
+
+    config = load_config()
+    if project in config:
+        raise ValueError("Project '{}' is already registered".format(project))
+
+    local_dir = input('Path to code_sync on this local machine: ')
+    target = input('Destination machine: ')
+    remote_dir = input('Path on the destination machine to sync: ')
+    port = int(input('Port number to use (default 22): ') or "22")
+
+    config_entry_data = {
+        project: {
```

Comment:
If somebody types an empty project at line 92, wouldn't the dictionary key fail? Would revert to a None, no?

Reply:
I realized that the argparse _requires_ an argument for the `--register` argument, you couldn't get into `register_project()` without supplying a project name on the command line. So I deleted the lines that checked for a `None` project and asked for one.